sdk/python/feast/infra/provider.py

```diff
@@ -0,0 +1,49 @@
+import abc
+from typing import List
+
+from feast import FeatureTable
+from feast.repo_config import RepoConfig
+
+
+class Provider(abc.ABC):
+    @abc.abstractmethod
+    def update_infra(
+        self,
+        project: str,
+        tables_to_delete: List[FeatureTable],
+        tables_to_keep: List[FeatureTable],
+    ):
+        """
+        Reconcile cloud resources with the objects declared in the feature repo.
+
+        Args:
+            tables_to_delete: Tables that were deleted from the feature repo, so provider needs to
+                clean up the corresponding cloud resources.
+            tables_to_keep: Tables that are still in the feature repo. Depending on implementation,
+                provider may or may not need to update the corresponding resources.
+        """
+        ...
+
+    @abc.abstractmethod
+    def teardown_infra(self, project: str, tables: List[FeatureTable]):
+        """
+        Tear down all cloud resources for a repo.
+
+        Args:
+            tables: Tables that are declared in the feature repo.
+        """
+        ...
+
+
+def get_provider(config: RepoConfig) -> Provider:
+    if config.provider == "gcp":
+        from feast.infra.gcp import Gcp
+
+        return Gcp(config.online_store.datastore)
+    elif config.provider == "local":
+        from feast.infra.local_sqlite import LocalSqlite
+
+        assert config.online_store.local is not None
```

Comment:
What would set `config.online_store.local`? Is it a default value?

Reply:
it is read from the config, there is no default right now but we can add that

Reply:
I think we need to have conventions for `local` and `gcp`. That way `assert` should never actually be hit, otherwise we should probably use a normal exception here.

Reply:
This assert is not supposed to be hit, it is there to make mypy happy. Config should be validated when it is being parsed, but due to how parsing works rn this member has to be `Optional`, so I need a type assertion here for mypy

Reply:
Ok, noted.